### PR TITLE
Set `Creating` and `Deleting` conditions close to `Status().Update()` calls 

### LINKED ISSUE #285: Status updates in `ExternalClient.Create` are omitted
### What problem are you facing?

With https://github.com/crossplane/crossplane-runtime/pull/283, we issue a standard update in all cases and that causes the `status` to be lost since managed resources treat it as subresource. But there are cases where some information about the resource is available only in the response of the creation call to the API.

### How could Crossplane help solve your problem?

We could possibly expand `resource.Managed` interface to have `GetStatus() runtime.Object` and `SetStatus(runtime.Object)` and then `DeepCopy` the status before the update operation and later call `SetStatus` on the returned object and issue a status update this time.

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Relates to #285

This approach causes us to repeat ourselves a bit, but prevents issues like the above, where refactoring caused us to accidentally overwrite a pending status update that we hadn't committed.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

I haven't yet.

### CODE REVIEW

File: pkg/reconciler/managed/reconciler.go [hunk 1]
```diff
@@ -607,7 +606,7 @@ func (r *Reconciler) Reconcile(_ context.Context, req reconcile.Request) (reconc
 			// condition. If not, we requeue explicitly, which will trigger
 			// backoff.
 			log.Debug("Cannot remove managed resource finalizer", "error", err)
-			managed.SetConditions(xpv1.ReconcileError(err))
+			managed.SetConditions(xpv1.Deleting(), xpv1.ReconcileError(err))
 			return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 		}
```
Comment (muvaf) [inline pkg/reconciler/managed/reconciler.go:612]: I think we need to set `xpv1.Creating()` in line 638 as well, which is called `if meta.ExternalCreateIncomplete(managed)`

Comment (muvaf) [reply pkg/reconciler/managed/reconciler.go:612]: Line 699 as well, in `if !observation.ResourceExists && meta.ExternalCreateSucceededDuring(managed, r.creationGracePeriod)`

Comment (muvaf) [reply pkg/reconciler/managed/reconciler.go:612]: Line 793, where `if err := r.client.Update(ctx, managed); err != nil`

**Review by muvaf — APPROVED** [review]
Need a few additions. The rest looks good to me but I think this fixes only one part of #285 . We still lose the status set in `external.Create`.

**Comment by negz** [comment]
> We still lose the status set in external.Create so we shouldn't close that issue.

Oh - I missed that. I didn't think we actually had any managed resources that modified status during `Create` (apart from setting the `Creating` status condition, which we now handle in the reconciler). Do you have an example of resources that do this? Now that I think about it I'm vaguely remembering some of the Azure resources recording operations there?

**Comment by muvaf** [comment]
I had not thought about Azure but yes and in fact, it could be showstopper since [the creation process](https://github.com/crossplane/provider-azure/blob/master/pkg/controller/database/mysqlserver/managed.go#L167) is being tracked with that ID. I was thinking about AWS, where the response of `Create` operation is [stored](https://github.com/crossplane/provider-aws/blob/master/pkg/controller/rds/dbinstance/zz_controller.go#L117) in status, which in some cases is different than `Describe`, though I couldn't find an example now.

FWIW, there is this [KEP](https://github.com/kubernetes/enhancements/pull/2537) by Tim Hockin that removes the litmus test of `If I erased the whole status struct, would everything in it be reconstructed (or at least be reconstructible) from observation?` So, `status` won't be treated as ephemeral as it used to be.

**Pull request merged**